docs/architecture/adr-013-proto-buf-serde.md

```diff
@@ -0,0 +1,136 @@
+[protoc-gen-tonic]: https://github.com/neoeinstein/protoc-gen-prost/blob/main/protoc-gen-tonic/README.md
+
+If the plugins are found lacking necessary features, try an [alternative approach][alt-gen]
+with `buf build --as-file-descriptor-set` and feeding the output to`tonic-build`. The serde
+implementations would need to be custom-derived like with the current approach.
+
+[alt-gen]: https://github.com/cometbft/cometbft-rs/issues/3#issuecomment-1916766090
+
+### API changes for protobuf/gRPC users
+
+The prost and tonic plugins invoke `prost-build` and `tonic-build`, respectively,
+so the code produced by them should not principally differ from the output of the
+current proto-compiler tool.
+
+### Changes in serialization
+
+The prost-serde plugin produces code depending on `pbjson`. This is
+an implementation detail not affecting the public API of the generated code,
+which exposes the regular serde trait implementations. The serialization schema
+produced by `pbjson-build` is compliant with the [JSON mapping for Protobuf][proto-json],
+to which the marshalling used by `encoding/json` in Go is mostly equivalent;
+notably, though, the JSON serialization currently used in CometBFT is implemented
+differently.
+
+[proto-json]: https://protobuf.dev/programming-guides/proto3/#json
+
+Our attempt to use the `protoc-gen-prost-serde` plugin will aim to provide serialization
+for most of the types generated from proto files; exceptions will be treated with
+setting custom attributes to `prost-build` via the buf plugin, much like in the
+current approach. A current limitation is that the `exclude` option of
+`pbjson-build`, needed to enable such overrides, is [not yet supported][protoc-gen-prost#84]
+by the buf plugin.
+
+[protoc-gen-prost#84]: https://github.com/neoeinstein/protoc-gen-prost/issues/84
+
+### Well-known types
+
+Where the CometBFT proto files make use of the well-known message types from
+the `google.protobuf` package, equivalent Rust types need to be substituted
+in the code generated for `cometbft-proto`. Developers of downstream crates
+may be interested in converging on the same types to avoid duplication.
+The CometBFT protobufs only use the `Duration` and `Timestamp` WKTs; protobufs
+in IBC, Cosmos-SDK, Interchain Security, and NFT transfer, also use `Any`.
+
+Three alternative WKT mapping approaches have been used across the protocol stack:
+
+1. [prost-types]. This is the most common mapping used by code generated with `prost-build`.
+   The crate does not provide serde implementations for the WKTs, so any message structs
+   including fields of WKTs either need overrides for their serde implementation or must omit
+   serde altogether.
+2. [pbjson-types]. The serde implementations provided by this WKT mapping crate is compliant
+   with the standard [protobuf JSON mapping][proto-json]. However, the JSON schema used by
+   CometBFT does not use the same format for `Duration`, so the crate cannot be substituted
```

Comment:
That seems to rule out this option completely, doesn't it? Or do you see a way to tweak the JSON schema to match what Comet is expecting?

Reply:
No. I've also looked at `proto-wkt-types` and they are no better in this regard (besides doing some custom datetime stuff, talk about mixing concerns).

I'm now leaning towards option 3 (keep the homespun `Timestamp`/`Duration` types) with `From` conversions between these and their most popular workalikes from other WKT crates. So the type disagreement can be solved by just throwing in an `.into()` instead of writing tedious member-by-member copies.

Reply:
> I'm now leaning towards option 3 (keep the homespun Timestamp/Duration types in cometbft-proto) with From conversions between these and their most popular workalikes from other WKT crates. So the type disagreement can be solved by just throwing in an .into() instead of writing tedious member-by-member copies.

Agreed, sounds like the best solution indeed.